packages/react-components/react-tree/src/hooks/useHTMLElementWalker.ts

```diff
@@ -0,0 +1,101 @@
+import { isHTMLElement } from '@fluentui/react-utilities';
+import * as React from 'react';
+import { assertIsDefinedRef } from '../utils/assert';
+
+export interface HTMLElementWalker {
+  readonly root: HTMLElement;
+  currentElement: HTMLElement;
+  firstChild(filter?: HTMLElementFilter): HTMLElement | null;
+  lastChild(filter?: HTMLElementFilter): HTMLElement | null;
+  nextElement(filter?: HTMLElementFilter): HTMLElement | null;
+  nextSibling(filter?: HTMLElementFilter): HTMLElement | null;
+  parentElement(filter?: HTMLElementFilter): HTMLElement | null;
+  previousElement(filter?: HTMLElementFilter): HTMLElement | null;
+  previousSibling(filter?: HTMLElementFilter): HTMLElement | null;
+}
+
+export type HTMLElementFilter = (element: HTMLElement) => number;
+
+const rootRefErrorMessage = 'useHTMLElementWalker: root reference not properly provided';
+
+export function useHTMLElementWalker(filter?: HTMLElementFilter) {
+  const rootRef = React.useRef<HTMLElement>(null);
```

Comment:
might want to create a callback ref where the rootRef is assigned, so you don't need to worry about the ref being correctly updated before/after render